### ISSUE TITLE
Set<Path> comparison no longer works as expected

### ISSUE THREAD

**Issue by AarjavP** [kotest/kotest#3032]
``` kotlin
    @Test
    fun `compare set of paths`() {
        val one = Paths.get("foo")
        val two = Paths.get("bar")
        val lhs = setOf(one, two)
        val rhs = setOf(one, two)
        lhs shouldBe rhs
    }
```
Using `io.kotest:kotest-assertions-core-jvm:5.0.3` this passes.

In version `5.1.0` and onwards, it fails with the message 
```
java.lang.AssertionError: Disallowed nesting iterator bar (WindowsPath) within bar (WindowsPath); (use custom test code instead)
```
(The message is constructed [here](https://github.com/kotest/kotest/blob/v5.3.0/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/IterableEq.kt#L154))
Note that `Path` implements `Iterable<Path>`. 
I do not know what the proper fix would be as I can definitely see the need to error out for custom Iterable types. Is there any additional arguments that be passed to say compare the elements of the set using the equals method?

**Comment by sksamuel** [comment]
Are you looking for something like `lhs should be equal using .equals rhs` ?

**Comment by AarjavP** [comment]
You're right, I could be doing `lhs == rhs shouldBe true` but Ideally I should be using something like `lhs shouldContainExactly rhs` so that kotest can print on the set differences (missing or extra expected elements)
I will try out that function when I get on again but I think it had the same issue.

since `Path` is a jdk type I would like it to work out of the box, but if its not possible or a good idea then maybe we can have something like `lhs.shouldContainExactly(rhs, Path::equals)`?

**Comment by sksamuel** [comment]
I think we should consider adding a `lhr shouldEqual rhs` which just uses .equals.

**Comment by tbvh** [comment]
I've been tinkering with an alternative. I don't know if it fits the architecture of kotest.

In IterableEq the method `nestedIterator` triggers the failure, `checkEquality` on Iterables doesn't want to operate on nested iterables. So, either nestedIterator could throw an exception that is caught so that `equals(Iterable, Iterable)` can fall back on `.equals`, or nested Iterators should be identified before going into the method.

For some[1] nested Iterators this can be checked using reflection, `IterableEq.isValidIterable(it: Any)` can make a jvm specific call to:


    private val nothingType = typeOf<List<Nothing>>().arguments.first().type!!

    private fun isNestedIterable(self: Any): Boolean {
       self is Iterable<*> || return false
       val superIterableKType = self::class.allSupertypes
          .find { it.classifier == Iterable::class } ?: return false
       val iteratedType = superIterableKType.arguments.first().type ?: return false
       if(iteratedType == nothingType) return false // emptyXXX() have Nothing type, but they can never be nested

       val selfKType = self::class.starProjectedType
       return (selfKType.isSubtypeOf(iteratedType) || selfKType.isSupertypeOf(iteratedType))
    }

[1] What can be checked are types that implement Iterable with type arguments matching itself, such as `class MyClass : Iterable<MyClass>`. I have a feeling that some configurations might slip through, but I haven't found it yet.

**Comment by AarjavP** [comment]
> I think we should consider adding a `lhr shouldEqual rhs` which just uses .equals.

I think it makes sense to have such method but in this case where I needed to compare two collections of items, I would miss the nice assertion message stating the differing elements

I do like tbvh's suggestion of adding additional checks for self type Iterable using reflection but I agree that it may still not cover all cases. I wonder of we can just have a simple config where we explicitly specify `Iterable<T>`s for which we need to use `.equals`. And on the JVM we can add `Path` to it by default.

Let me know how if there's anything I can contribute in terms of code. I can an attempt `shouldEqual`

**Comment by Kantis** [comment]
Related to https://github.com/kotest/kotest/issues/3130

I think fixing #3130 as discussed there (dropping IterableEq assumption) would also resolve this as well

**Comment by sksamuel** [comment]
Fixed in 5.6.1
